src/model/markercollection.js

```diff
@@ -0,0 +1,287 @@
+	}
+
+	/**
+	 * Creates and adds a {@link ~Marker marker} to the `MarkerCollection` with given name on given
+	 * {@link module:engine/model/range~Range range}.
+	 *
+	 * If `MarkerCollection` already had a marker with given name (or {@link ~Marker marker} was passed) and the range to
+	 * set is different, the marker in collection is removed and then new marker is added. If the range was same, nothing
+	 * happens and `false` is returned.
+	 *
+	 * @fires {module:engine/model/markercollection~MarkerCollection#event:add}
+	 * @fires {module:engine/model/markercollection~MarkerCollection#event:remove}
+	 * @param {String|~Marker} markerOrName Name of marker to add or Marker instance to update.
+	 * @param {module:engine/model/range~Range} range Marker range.
+	 * @returns {~Marker} `Marker` instance added to the collection.
+	 */
+	set( markerOrName, range ) {
+		const markerName = markerOrName instanceof Marker ? markerOrName.name : markerOrName;
+
+		if ( this._markers.has( markerName ) ) {
+			this.remove( markerName );
+		}
+
+		const liveRange = LiveRange.createFromRange( range );
+		const marker = new Marker( markerName, liveRange );
+
+		this._markers.set( markerName, marker );
+		this.fire( 'add', marker );
+
+		return marker;
+	}
+
+	/**
+	 * Removes given {@link ~Marker marker} or a marker with given name from the `MarkerCollection`.
+	 *
+	 * @param {String} markerOrName Marker or name of a marker to remove.
+	 * @returns {Boolean} `true` if marker was found and removed, `false` otherwise.
+	 */
+	remove( markerOrName ) {
+		const markerName = markerOrName instanceof Marker ? markerOrName.name : markerOrName;
+		const marker = this._markers.get( markerName );
+
+		if ( marker ) {
+			this._markers.delete( markerName );
+			this.fire( 'remove', marker );
+
+			this._destroyMarker( marker );
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Destroys markers collection.
+	 */
+	destroy() {
+		for ( let marker of this._markers.values() ) {
+			this._destroyMarker( marker );
+		}
+
+		this._markers = null;
+
+		this.stopListening();
+	}
+
+	/**
+	 * Destroys marker.
+	 *
+	 * @private
+	 * @param {~Marker} marker Marker to destroy.
+	 */
+	_destroyMarker( marker ) {
+		marker.stopListening();
+		marker._liveRange.detach();
+		marker._liveRange = null;
+	}
+
+	/**
+	 * Fired whenever marker is added to `MarkerCollection`.
+	 *
+	 * @event add
+	 * @param {~Marker} The added marker.
+	 */
+
+	/**
+	 * Fired whenever marker is removed from `MarkerCollection`.
+	 *
+	 * @event remove
+	 * @param {~Marker} marker The removed marker.
+	 */
+}
+
+mix( MarkerCollection, EmitterMixin );
+
+/**
+ * `Marker` is a continuous parts of model (like a range), is named and represent some kind of information about marked
+ * part of model document. In contrary to {@link module:engine/model/node~Node nodes}, which are bits of data, markers are
+ * kind of bookmarks, marking part of a model document between marker start and marker end.
```

Comment:
If you are a CKE4 developer reading this doc, you can think that you should use markers for cases you used bookmarks in CKE4, what is wrong. You should use live ranges for these cases.